harness/determined/experimental/core_v2.py

```diff
@@ -0,0 +1,219 @@
+checkpoint = None  # type: Optional[core.CheckpointContext]
+searcher = None  # type: Optional[core.SearcherContext]
+info = None  # type: Optional[determined.ClusterInfo]
+
+
+@dataclasses.dataclass
+class DefaultConfig:
+    """
+    In the future, `DefaultConfig` options will be merged with the experiment config values when
+    running in the managed mode.
+    """
+
+    name: Optional[str] = None
+    hparams: Optional[Dict[str, Any]] = None
+    data: Optional[Dict] = None
+    description: Optional[str] = None
+    labels: Optional[List[str]] = None
+    # Also to be added:
+    # - hyperparameters: const only
+    # - checkpoint_policy: for optional gc
+    # For managed mode, workspace and project MUST be present in the exp conf for RBAC reasons.
+    # We expose them as separate arguments on the `det.init` call.
+    # - workspace
+    # - project
+    # Unsupported:
+    # - bind_mounts
+    # - data_layer
+    # - debug
+    # - entrypoint
+    # - environment
+    # - internal
+    # - max_restarts
+    # - optimizations
+    # - profiling
+    # - reproducibility
+    # - security
+    # - slurm
+    # Searcher:
+    # - searcher
+    # - min_checkpoint_period
+    # - min_validation_period
+    # - pbs
+    # - perform_initial_validation
+    # - records_per_epoch
+    # - scheduling_unit
+    # Deprecated:
+    # - data_layer
+    # - pbs
+    # - tensorboard_storage
+    checkpoint_storage: Optional[Union[str, Dict[str, Any]]] = None
+    # Searcher is currently a hack to disambiguate single trial experiments
+    # and hp searches in WebUI.
+    # Also searcher metric is useful for sorting in the UI and, in the future, checkpoint gc.
+    searcher: Optional[Dict[str, Any]] = None
+    # TODO(ilia): later to be replaced with:
+    # Unmanaged mode only config options:
+    # multi_trial_experiment: bool = False
+    # metric: Optional[str] = None
+    # smaller_is_better: bool = True  # mode?
+
+
+@dataclasses.dataclass
+class UnmanagedConfig:
+    """
+    `UnmanagedConfig` values are only used in the unmanaged mode.
+    """
+
+    # For the managed mode, workspace is critical for RBAC so it cannot be easily
+    # merged and patched during the experiment runtime.
+    workspace: Optional[str] = None
+    project: Optional[str] = None
+    # External experiment & trial ids.
+    experiment_id: Optional[Union[str, int]] = None
+    trial_id: Optional[Union[str, int]] = None
+
+
+def _set_globals() -> None:
+    global train
+    global checkpoint
+    global distributed
+    global preempt
+    global searcher
+    global info
+
+    assert _context is not None
+    train = _context.train
+    checkpoint = _context.checkpoint
+    distributed = _context.distributed
+    preempt = _context.preempt
+    searcher = _context.searcher
+    info = _context.info
+
+
+def init(
+    *,
+    defaults: Optional[DefaultConfig] = None,
+    unmanaged: Optional[UnmanagedConfig] = None,
+    client: Optional[Determined] = None,
```

Comment:
it seems odd to me that we're passing in the client. are we thinking about moving this `init` method to belong to the client in the future?

Reply:
we have the singleton client (`determined.experimental.client` module) but you can also instantiate a client instance (`determined.common.experimental.determined.Determined` class) — I believe we need a way to enable using these.
I don't really see a case for having this `init` be a method on `Determined` class itself.

Reply:
fwiw, in my mind, I was thinking to enable this exact functionality in `core.init()` by accepting master_url, user, password, etc in the `core.init()` function.

I don't have an opinion on what is better, but I wanted to list it as a possible option.